src/main/kotlin/com/impactupgrade/integration/hubspot/crm/v3/PropertiesCrmV3Client.kt

```diff
@@ -0,0 +1,157 @@
+        .build()
+    val response: HttpResponse<String> = HttpClient.newBuilder().build().send(request, HttpResponse.BodyHandlers.ofString())
+
+    return when (response.statusCode()) {
+      200 -> {
+        val responseEntity = objectMapper.readValue(response.body(), Property::class.java)
+        log.info("HubSpot API response {}: {}", response.statusCode(), responseEntity)
+        responseEntity
+      }
+      else -> {
+        val retryFunction = { newAttemptCount: Int -> update(objectType, propertyName, property, newAttemptCount) }
+        handleError(response, attemptCount, retryFunction)
+      }
+    }
+  }
+
+  fun delete(objectType: String, propertyName: String) {
+    log.info("deleting {}'s property {}", objectType, propertyName)
+    val response = target
+      .path(objectType)
+      .path(propertyName)
+      .queryParam("hapikey", apiKey)
+      .request(MediaType.APPLICATION_JSON)
+      .delete()
+    log.info("HubSpot API response: {}", response.status)
+  }
+
+  // for Java callers
+  fun readAll(objectType: String) = readAll(objectType, 0)
+
+  fun readAll(objectType: String, attemptCount: Int = 0): PropertiesResponse? {
+    log.info("fetching {} properties:", objectType)
+
+    val response = target
+      .path(objectType)
+      .queryParam("hapikey", apiKey)
+      .request(MediaType.APPLICATION_JSON)
+      .get()
+    return when (response.status) {
+      200 -> {
+        val responseEntity = response.readEntity(PropertiesResponse::class.java)
+        log.info("HubSpot API response {}: {}", response.status, responseEntity)
+        responseEntity
+      }
+      else -> {
+        val retryFunction = { newAttemptCount: Int -> readAll(objectType, newAttemptCount) }
+        handleError(response, attemptCount, retryFunction)
+      }
+    }
+  }
+
+  // for Java callers
+  fun batchInsert(objectType: String, properties: List<Property>) = batchInsert(objectType, properties, 0)
+
+  fun batchInsert(objectType: String, properties: List<Property>? = null, attemptCount: Int = 0): PropertiesResponse? {
+    log.info("inserting properties: {}", properties)
+    val batchRequest = PropertyBatchCreateRequest(properties)
+    val response = target
+      .path(objectType + "/batch/create")
+      .queryParam("hapikey", apiKey)
+      .request(MediaType.APPLICATION_JSON)
+      .post(Entity.entity(batchRequest, MediaType.APPLICATION_JSON_TYPE))
+    return when (response.status) {
+      201 -> {
+        val responseEntity = response.readEntity(PropertiesResponse::class.java)
+        log.info("HubSpot API response {}: {}", response.status, responseEntity)
+        responseEntity
+      }
+      else -> {
+        val retryFunction = { newAttemptCount: Int -> batchInsert(objectType, properties, newAttemptCount) }
+        handleError(response, attemptCount, retryFunction)
+      }
+    }
+  }
+
+  //TODO: property groups?
```

Comment:
Might be helpful in the future, but we're typically inserting properties into existing groups...

Reply:
Removed the todo